### ISSUE TITLE
Segfault after FileIndex:Indexing on headless server

### ISSUE THREAD

**Issue by DotStarMoney** [OpenRCT2/OpenRCT2#10475]
**OS:** Ubuntu 19.04
**Commit/Build:** 0.2.4-develop-0~773-201912291905-ge36d87e~ubuntu19.04.1

Attempting to start a headless server on Ubuntu 19.04 using the above build starts to build the object index, logs `VERBOSE: FileIndex:Indexing '/usr/share/openrct2/object/official/station/openrct2.station.noentrance.json'` then segfaults. It sometimes will follow the aforementioned line with `VERBOSE: FileIndex:Indexing '/usr/share/openrct2/object/rct2/ride/rct2.fsauc.json'` before the crash.

- [ ] Reproducible in RCT2 (vanilla)?
- [ ] Specific to multiplayer?

**Steps to reproduce:**
1. `openrct2-cli host "my_save_path.sv6" --verbose`

**Comment by IntelOrca** [comment]
Are you able to provide a stack trace?

**Comment by DotStarMoney** [comment]
My version doesn't have debug symbols so this (valgrind) is the best I can do I think (so not really useful):


VERBOSE: FileIndex:Indexing '/usr/share/openrct2/object/official/station/openrct2.station.noentrance.json'
==1698== Thread 3:
==1698== Invalid read of size 8
==1698==    at 0x226D54: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x226680: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x287401: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x289123: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x283E8D: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x24EFF1: ??? (in /usr/bin/openrct2-cli)
==1698==    by 0x500662F: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.26)
==1698==    by 0x4857181: start_thread (pthread_create.c:486)
==1698==    by 0x5383B1E: clone (clone.S:95)
==1698==  Address 0x2b is not stack'd, malloc'd or (recently) free'd

**Comment by janisozaur** [comment]
This is not useful at all. Please install debug symbols and provide symbolicated stack trace

**Comment by DotStarMoney** [comment]
Gladly, is there a resource for how to do that?

**Comment by janisozaur** [comment]
https://wiki.ubuntu.com/DebuggingProgramCrash#Installing_dbgsym_packages_from_a_PPA

**Comment by MalteJoe** [comment]
I have the same issue although not running headless but with Ubuntu 19.04. I installed debug symbols as suggested and this is what the trace looks like:

```
==18838== Command: openrct2 --verbose
==18838== 
OpenRCT2,  ( on , DEBUG)
Linux (x86-64)
@ 

VERBOSE: time is 2020/01/04 14:39:31
VERBOSE: registering AVX2 mask function
VERBOSE: Looking for OpenRCT2 data in /usr/bin/data
[...]
Building object index (2101 items)
VERBOSE: FileIndex:Indexing '/usr/share/openrct2/object/official/station/openrct2.station.noentrance.json'
==18838== Thread 3:
==18838== Invalid read of size 8
==18838==    at 0x302874: std::vector<std::filesystem::__cxx11::path::_Cmpt, std::allocator<std::filesystem::__cxx11::path::_Cmpt> >::~vector() (stl_vector.h:567)
==18838==    by 0x3021A0: ~path (fs_path.h:208)
==18838==    by 0x3021A0: ~_Cmpt (fs_path.h:643)
==18838==    by 0x3021A0: _Destroy<std::filesystem::__cxx11::path::_Cmpt> (stl_construct.h:98)
==18838==    by 0x3021A0: __destroy<std::filesystem::__cxx11::path::_Cmpt*> (stl_construct.h:108)
==18838==    by 0x3021A0: _Destroy<std::filesystem::__cxx11::path::_Cmpt*> (stl_construct.h:137)
==18838==    by 0x3021A0: _Destroy<std::filesystem::__cxx11::path::_Cmpt*, std::filesystem::__cxx11::path::_Cmpt> (stl_construct.h:206)
==18838==    by 0x3021A0: ~vector (stl_vector.h:567)
==18838==    by 0x3021A0: ~path (fs_path.h:208)
==18838==    by 0x3021A0: Path::GetExtension(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (Path.cpp:147)
==18838==    by 0x3749A1: ObjectFileIndex::Create(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const (ObjectRepository.cpp:95)
==18838==    by 0x3766C3: FileIndex<ObjectRepositoryItem>::BuildRange(int, FileIndex<ObjectRepositoryItem>::ScanResult const&, unsigned long, unsigned long, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<unsigned long>&, std::mutex&) const (FileIndex.hpp:188)
==18838==    by 0x37142D: __invoke_impl<void, void (FileIndex<ObjectRepositoryItem>::*&)(int, const FileIndex<ObjectRepositoryItem>::ScanResult&, long unsigned int, long unsigned int, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<long unsigned int>&, std::mutex&) const, const FileIndex<ObjectRepositoryItem>*&, int&, const FileIndex<ObjectRepositoryItem>::ScanResult&, long unsigned int&, long unsigned int&, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<long unsigned int>&, std::mutex&> (invoke.h:73)
==18838==    by 0x37142D: __invoke<void (FileIndex<ObjectRepositoryItem>::*&)(int, const FileIndex<ObjectRepositoryItem>::ScanResult&, long unsigned int, long unsigned int, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<long unsigned int>&, std::mutex&) const, const FileIndex<ObjectRepositoryItem>*&, int&, const FileIndex<ObjectRepositoryItem>::ScanResult&, long unsigned int&, long unsigned int&, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<long unsigned int>&, std::mutex&> (invoke.h:95)
==18838==    by 0x37142D: __call<void, 0, 1, 2, 3, 4, 5, 6, 7> (functional:400)
==18838==    by 0x37142D: operator()<> (functional:484)
==18838==    by 0x37142D: std::_Function_handler<void (), std::_Bind<void (FileIndex<ObjectRepositoryItem>::*(FileIndex<ObjectRepositoryItem> const*, int, std::reference_wrapper<FileIndex<ObjectRepositoryItem>::ScanResult const>, unsigned long, unsigned long, std::reference_wrapper<std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> > >, std::reference_wrapper<std::atomic<unsigned long> >, std::reference_wrapper<std::mutex>))(int, FileIndex<ObjectRepositoryItem>::ScanResult const&, unsigned long, unsigned long, std::vector<ObjectRepositoryItem, std::allocator<ObjectRepositoryItem> >&, std::atomic<unsigned long>&, std::mutex&) const> >::_M_invoke(std::_Any_data const&) (std_function.h:297)
==18838==    by 0x32E901: operator() (std_function.h:687)
==18838==    by 0x32E901: JobPool::ProcessQueue() (JobPool.hpp:144)
==18838==    by 0x539462F: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.26)
==18838==    by 0x4FC0181: start_thread (pthread_create.c:486)
==18838==    by 0x5711B1E: clone (clone.S:95)
==18838==  Address 0x2b is not stack'd, malloc'd or (recently) free'd
```

As I am not using the cli executable, that's why the memory addresses don't match. I'm on version `0.2.4-develop-0~795-202001021905-ge87607f~ubuntu19.04.1` of the Ubuntu PPA

**Comment by MalteJoe** [comment]
I have traced down the error and this looks like it is an Ubuntu specific issue with gcc8 and `std::filesystem` support as the comments here suggest: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90050

Either upgrading to gcc9 or providing the linker flag `-lstdc++fs` should therefore fix this.
I just tried compiling with gcc9 and it worked like a charm.

I'd love to create a PR but I could use some hints where to configure this stuff to not break other builds.
